Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP1StateMachine.swift

```diff
@@ -519,16 +519,20 @@ extension HTTPConnectionPool {
         // MARK: HTTP2
 
         mutating func newHTTP2MaxConcurrentStreamsReceived(_ connectionID: Connection.ID, newMaxStreams: Int) -> Action {
-            // It is save to bang the http2Connections here. If we get this callback but we don't have
-            // http2 connections something has gone terribly wrong.
-            _ = self.http2Connections!.newHTTP2MaxConcurrentStreamsReceived(connectionID, newMaxStreams: newMaxStreams)
+            // The `http2Connections` are optional here:
+            // Connections report events back to us, if they are in a shutdown that was
+            // initiated by the state machine. For this reason this callback might be invoked
```

Comment:
This doesn't quite make sense to me. I think you're saying that we can still receive events after shutdown was initiated by the state machine, and when that happens `http2Connections` will be `nil` and we can safely ignore the event?

Reply:
I think it'd be worth adding a comment where `http2Connections` is declared on L31 saying when we expect it to be `nil`.

Reply:
I also wonder if `http2Connections` ever needs to be `nil` -- presumably we can just remove all connections from it?

Reply:
I think the only reason why we made it optional was to free some memory early. I think it would make the code clearer if we make it non-optional e.g.: after we have closed a connection we set it to nil if it is empty.
https://github.com/swift-server/async-http-client/blob/12130d30cc5f5a01a48143b1a11f6756cb3ee5f6/Sources/AsyncHTTPClient/ConnectionPool/State%20Machine/HTTPConnectionPool%2BHTTP1StateMachine.swift#L543-L545
Instead we could just check if it is empty in all places where we currently check that it is nil e.g. during shutdown and make `http2Connections` non-optional.

The same is true in the `HTTP2StateMachine` for `http1Connections`.

Reply:
@dnadoba will fix this in a follow up PR.